File: android/app/src/main/java/dev/rahier/pouleparty/ui/gamelogic/ChallengeProgress.kt
package dev.rahier.pouleparty.ui.gamelogic

import dev.rahier.pouleparty.model.Challenge
import dev.rahier.pouleparty.model.ChallengeType
import kotlin.math.ceil

data class LevelProgress(val validated: Int, val total: Int, val threshold: Int)

/**
 * Pure helpers mirroring iOS `ChallengeProgress`. A hunter unlocks
 * `level N+1` once they have validated at least `ceil(N_oneShot × 0.80)`
 * of the oneShot challenges in `level N`. Repeatable challenges are
 * unbounded by design and are excluded from the calculation.
 */
object ChallengeProgress {

    fun isLevelUnlocked(
        level: Int,
        challenges: List<Challenge>,
        validatedChallengeIds: Set<String>,
    ): Boolean {
        if (level <= 1) return true
        val progress = levelProgress(level, challenges, validatedChallengeIds)
        return progress.validated >= progress.threshold
    }

    fun levelProgress(
        level: Int,
        challenges: List<Challenge>,
        validatedChallengeIds: Set<String>,
    ): LevelProgress {
        val previousLevel = level - 1
        val oneShotsAtPrevious = challenges.filter {
            it.level == previousLevel && it.typeEnum == ChallengeType.ONE_SHOT
        }
        val total = oneShotsAtPrevious.size
        val validated = oneShotsAtPrevious.count { validatedChallengeIds.contains(it.id) }
        val threshold = ceil(total * 0.80).toInt()
        return LevelProgress(validated = validated, total = total, threshold = threshold)
    }
}

File: ios/PouleParty/Components/GameLogic/ChallengeProgress.swift
//
//  ChallengeProgress.swift
//  PouleParty
//

import Foundation

enum ChallengeProgress {
    /// Returns `true` when the hunter has validated at least
    /// `ceil(N × 0.80)` of the `oneShot` challenges in `level - 1`,
    /// where N is the total oneShot count at that previous level.
    /// `level == 1` is always unlocked. Repeatable challenges are
    /// excluded from the calculation: they have no upper bound so
    /// they can never gate progression.
    static func isLevelUnlocked(
        level: Int,
        challenges: [Challenge],
        validatedChallengeIds: Set<String>
    ) -> Bool {
        guard level > 1 else { return true }
        let progress = levelProgress(
            level: level,
            challenges: challenges,
            validatedChallengeIds: validatedChallengeIds
        )
        return progress.validated >= progress.threshold
    }

    /// Returns the count of oneShot challenges already validated for
    /// `level - 1`, the total oneShot count at that previous level,
    /// and the `ceil(N × 0.80)` threshold required to unlock `level`.
    /// The UI uses this to render the "🔒 Unlock at X / Y" label.
    /// When `level == 1` (or `level - 1` has no oneShot challenges)
    /// every component returns 0, which keeps the level unlocked.
    static func levelProgress(
        level: Int,
        challenges: [Challenge],
        validatedChallengeIds: Set<String>
    ) -> (validated: Int, total: Int, threshold: Int) {
        let previousLevel = level - 1
        let oneShotsAtPrevious = challenges.filter {
            $0.level == previousLevel && $0.type == .oneShot
        }
        let total = oneShotsAtPrevious.count
        let validated = oneShotsAtPrevious.filter {
            validatedChallengeIds.contains($0.id)
        }.count
        let threshold = Int(ceil(Double(total) * 0.80))
        return (validated, total, threshold)
    }
}
